### PR TITLE
Add support for CSI 18t 

### LINKED ISSUE #13944: CSI 18t "Report Size In Characters" not working.
### Windows Terminal version

1.15.2282.0

### Windows build number

10.0.19044.1889

### Other Software

Own project.

### Steps to reproduce

Start CMD or PS within Terminal.
Run app which activates VT-Modes for IN and OUT.
Send **CSI 18 t** to stdout.
Read (wait for) response on stdin.

### Expected Behavior

A response from the Terminal a la:
**CSI** 'Decimal-Rows' **;** 'Decimal-Columns' **t**

### Actual Behavior

No response from the Terminal.
But i.e. sending **CSI 6 n** (Report Actual Cursor Position) results as expected in a response a la:
**CSI** 'Decimal-Row' **;** 'Decimal-Column' **R**

### PR DESCRIPTION
**Pull request by michalnpl**
Adds support for CSI 18t to report the buffer screen size in characters.

This pull request adds support for **CSI 18t**. When submitted to the terminal, it will respond with **"\033[8;{A};{B}t"** where **A** is equal to the **height** and **B** is equal to the **width** of the screen buffer in the number of characters (not pixels).

## Validation Steps Performed
Manual tests against PowerShell 7 and ConHost.
Added adapterTest

Closes #13944

### CODE REVIEW

**Comment by michalnpl** [comment]
> @microsoft-github-policy-service agree [company="{your company}"]

@microsoft-github-policy-service agree [company="{default}"]

**Comment by michalnpl** [comment]
> @michalnpl the command you issued was incorrect. Please try again.
> 
> Examples are:
> 
> ```
> @microsoft-github-policy-service agree
> ```
> 
> and
> 
> ```
> @microsoft-github-policy-service agree company="your company"
> ```

@microsoft-github-policy-service agree

**Comment by michalnpl** [comment]
There are no TAEF tests. If you can add tests to this PR, I would greatly appreciate it so I can learn.

File: src/terminal/adapter/adaptDispatch.cpp
Comment (zadjii-msft) [inline src/terminal/adapter/adaptDispatch.cpp:3073]: I'm tempted to say that this is okay, but IIRC @j4james thought we should use the buffer width, not just the viewport width here.

**Review by zadjii-msft — COMMENTED** [review]
FWIW, I think we'll be swamped getting 1.18 out the door, so reviews might be slow. 

You might be able to do something like

```c++
void ScreenBufferTests::TestGetWindowSizeInVt()
{
    auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
    auto& si = gci.GetActiveOutputBuffer();
    auto& stateMachine = si.GetStateMachine();
    auto& cursor = si.GetTextBuffer().GetCursor();

    stateMachine.ProcessString(L"\x1b[18t");
    
    
    gci.pInputBuffer.Read(...);
```

and then check the `Char` of each of those key events. 


Sorry I'm not more help ATM. Maybe after we sign off on 1.18 I can help more. We apparently _didn't_ have tests like this, so I don't blame you for not writing a test for this one. I don't think there _is_ prior art

**Comment by michalnpl** [comment]
@zadjii-msft @j4james I adressed:

> apps are more likely to expect the dimensions of the addressable cell range, which for us, is determined by the _buffer_ width combined with the _viewport_ height.

with and tested manually (still no TAEF):

image

@zadjii-msft 
No worries; maybe by the time you get to this one, I can figure out the tests 🙃

File: src/terminal/adapter/adaptDispatch.cpp
Comment (michalnpl) [reply src/terminal/adapter/adaptDispatch.cpp]: Adressed.

**Comment by j4james** [comment]
Tests for sequences like this are typically handled in [`AdapterTest`](https://github.com/microsoft/terminal/blob/main/src/terminal/adapter/ut_adapter/adapterTest.cpp). Just search for anything using `ValidateInputEvent`.

For example, this is a test case for the `DECRQSS` margin query, which first sets up the viewport height to the range it's expecting to see in the query response.

https://github.com/microsoft/terminal/blob/10c6206bfef7d045347327687266832931f47eb6/src/terminal/adapter/ut_adapter/adapterTest.cpp#L1624-L1630

That seems similar to the sort of thing you might want to do. And note that you can access the text buffer the same way as the viewport with `_testGetSet->_textBuffer`, although I'm not sure about resizing it. You could just assume a buffer width of 100 for your tests, since that's what the `PrepData` function initializes it with by default (see [here](https://github.com/microsoft/terminal/blob/10c6206bfef7d045347327687266832931f47eb6/src/terminal/adapter/ut_adapter/adapterTest.cpp#L283)).

**Comment by michalnpl** [comment]
I've added a TAEF test for this.

File: src/terminal/adapter/adaptDispatch.cpp
Comment (j4james) [inline src/terminal/adapter/adaptDispatch.cpp:3074]: Nit: I think that `std::wstring` should automatically cast to `std::wstring_view`, so you shouldn't need an explicit cast on the `textSize` parameter. But with that out the way, you might as well embed the `fmt::format` call directly within the `ReturnResponse` call and avoid the `textSize` variable altogether. I suspect we're doing the same thing in a bunch of other places, though, so it's not a big deal.

Comment (michalnpl) [reply src/terminal/adapter/adaptDispatch.cpp]: Addressed.

**Review by j4james — COMMENTED** [review]
I can't officially approve anything, but this looks good to me. Thanks.

**Review by j4james — APPROVED** [review]

**Review by zadjii-msft — APPROVED** [review]
Thanks for bearing with to write the test!

**Comment by zadjii-msft** [comment]
_Screenshots from body:_


For test code:
```
#include <stdio.h>

int main() {
    printf("\033[18t");
    return 0;
}
```

Ubuntu:
image

Terminal (both PowerShell and Conhost):
image

and for Ubuntu test code:
```
#include <stdio.h>
#include <unistd.h>
#include <termios.h>

int main() {
    struct termios old_termios, new_termios;
    tcgetattr(STDIN_FILENO, &old_termios);
    new_termios = old_termios;
    new_termios.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(STDIN_FILENO, TCSANOW, &new_termios);
    printf("\033[18t");
    int rows, cols;
    scanf("\033[8;%d;%dt", &rows, &cols);
    tcsetattr(STDIN_FILENO, TCSANOW, &old_termios);
    printf("Terminal size: %d columns x %d rows\n", cols, rows);
    return 0;
}
```
image

and for **microsoft/terminal** test code:
```
#include <stdio.h>
#include <windows.h>

int main() {
    HANDLE hConsole = GetStdHandle(STD_INPUT_HANDLE);
    DWORD prev_mode;
    GetConsoleMode(hConsole, &prev_mode);
    DWORD new_mode = prev_mode & ~(ENABLE_LINE_INPUT | ENABLE_ECHO_INPUT);
    SetConsoleMode(hConsole, new_mode);
    printf("\033[18t");
    int rows, cols;
    scanf_s("\033[8;%d;%dt", &rows, &cols);
    SetConsoleMode(hConsole, prev_mode);
    printf("Terminal size: %d columns x %d rows\n", cols, rows);
    return 0;
}
```
image

Kindly please advise if this is the desired behavior.

**Comment by michalnpl** [comment]
Going to need another approver to merge this, as @j4james mentioned he could not approve anything officially.

File: src/terminal/adapter/adaptDispatch.cpp
Comment (DHowett) [inline src/terminal/adapter/adaptDispatch.cpp:3062]: nit: all of the other uses of `fmt` in this file use `FMT_COMPILE` for the format string. That keeps the code size down when this is compiled into Windows' conhost.exe.

Would you mind doing the same? :smile:

**Review by DHowett — CHANGES_REQUESTED** [review]

**Review by DHowett — COMMENTED** [review]
Thanks!

**Review by DHowett — APPROVED** [review]
Am I going crazy, or did Refined GitHub change the order of the signoff buttons?

**Pull request merged**
